### ISSUE TITLE
Loot Generation In Dungeons Behaving Strangely

### ISSUE THREAD

**Issue by TeetoKaziim** [BlackJar72/DoomlikeDungeons#7]
**Summary of Issue**
- Loot from the "Loot" category only generates loot from "loot, 7", and doesn't generate any loot from 1, 2, 3, 4, 5, 6, or 8. (Actual issue, almost certainly not intended)
- Loot from the "Heal" and "Gear" categories higher than "heal, 4" and "gear, 4" does not seem to generate at all, or at least, at what seems like an unintentionally low rate. (Possible issue, seems odd)

**Overview:**
Playtesting my pack I'm working on, I noticed as I was clearing out dungeons that I was getting an abundance of some kinds of loot, and seeing almost none of other kinds. As the sample size grew, so did my suspicions, and I became convinced that certain loot types were either not generating, or generating at what can only be described as a wildly low rate. So I decided to do some science.

**Doing Of Science**
I whipped up a minecraft 1.12.2 instance with the following mods:

DoomlikeDungeons-1.10.1-MC1.12.2.jar
Draconic-Evolution-1.12-2.3.5.269-universal.jar (Included solely for its item magnet for this test)
BrandonsCore-1.12-2.3.4.134-universal.jar
CodeChickenLib-1.12-3.1.3.313-universal.jar
RedstoneFlux-1.12-2.0.1.2-universal.jar (These 3 are dependencies for draconic evolution, geez. RedstoneFlux is the only one of these that also appears in my "actual" pack I'm working on.)
Hwyla-1.8.23-B38_1.12.jar
jei_1.12.2-4.8.5.136.jar (Make testing more convenient)
forge-1.12.2-14.23.1.2562.jar (I might be behind on forge versions, I forgot to check before testing, so I'm sharing this if it ends up being relevant.)

**My Relevant Configuration Files:**

- DLDungeons.JBG.cfg: https://pastebin.com/raw/hVZTuGPK (I set easy-find on, though I don't know why)

- chests.cfg: https://pastebin.com/raw/H86K8NSG (Only 1 item in each tier of each category, with a stack size range of 1 to 1. Basically, the idea being I wanted each "item that ended up in my inventory" to represent "1 loot roll", so that my 5 stacks of poison potatoes could be directly compared with my 3 stacks of spider eyes. Also, the specific items are edited so that it's all things that stack up to 64)

- common.cfg (In my Themes folder, only theme there): https://pastebin.com/raw/bfjVb9gC (Notable: I set the only allowed dungeon size to huge, in case something about loot generation was tied to getting a big dungeon, and, to make the sample size larger with less work. Also notable: There is at least one mob in each spawner tier category, though, anecdotally, I don't think I saw any spawners from any of: bruteMobs, eliteMobs, or bossMobs, though I was not rigourous, just kinda keeping my eyes open as I flew around slapping chests.) 

Then, I loaded up a Creative Mode, Superflat World, with structures enabled, and Peaceful difficulty on, so that mobs would stay out of the way and not die and drop items, polluting my data. Grabbed a Draconic item magnet from JEI, then, flew around, using dldspawn to generate 4 (huge sized!) dungeons, and flew around in creative breaking the chests, and vacuuming up the loot. After breaking what I later realized is a large number of chests, but a number I should have been counting (sorry), I ended up with the following loot distribution:

My Results: https://i.imgur.com/TEASFh7.png
A text breakdown of my results because what kind of loser submits data in .png form:

**From the "gear" loot category:**
- 337 pistons (gear, 1)
- 131 hoppers (gear, 2)
- 50 cauldrons (gear, 3)
- 1 tnt (gear, 4)
- 0 items from any of (gear, 5) (gear, 6) (gear, 7)

**From the "heal" loot category:**
- 376 poisonous potatos (heal, 1)
- 228 spider eyes (heal, 2)
- 75 rotten flesh (heal, 3)
- 3 beetroots (heal, 4)
- 0 items from the following: (heal, 5) (heal, 6) (heal, 7)

**From the "loot" loot category:**
- 0 items from the following: (loot, 1) (loot, 2) (loot, 3) (loot, 4) (loot, 5) (loot, 6)
- 106 diamonds (loot, 7)
- 0 items from: (loot, 8)

**From (?????) loot category:**
- 5 enchanted books (?????) (I guess these still generate even if you don't put any books in your loot pools?)

Total: 1312 loot rolls collected.

**My own conclusions/suspicions:**

Maybe the higher tier loot is mean to be that rare, in the gear and heal categories, but maybe the numbers are tuned a bit smaller than intended, so I felt like it was worth sharing that data. Especially with how wide a swath of the upper tiers just don't show up at all, and with the kind of loot that seems to occupy those tiers in the default chests.cfg file, I felt like maybe you might not have intended them to be quite THAT rare.

The real weirdo anomaly, which also showed up in my testing of my actual modpack with real loot pools, is the behaviour of the "loot" category, and this is what tipped me off that something was amiss. I'm almost certain you don't mean for ONLY tier 7 items to spawn.

**Denouement**
Thank you for your time, I know this is a fairly large pile of text to go through, and I don't really mean to heap work onto your plate unless you have actual enthusiasm for it. I know I've been a mosquito in your ear for the past couple weeks, so I just wanna say thank you very much, and that I hope you know it's my own enthusiasm and enjoyment of this mod that drives me. I've had a lot of fun with it.

**Comment by TeetoKaziim** [comment]
So did some followup testing, adding a
"Flags =  HARD" to my theme and kicking the DLDungeonsJBG.cfg difficulty up to "5: nightmare". I ran into some of the higher spawners, (even saw a witch, which is currently in the "boss" slot), but, still saw a similar lack of "gear" and "heal" beyond tier 4.

A somewhat tangential question, but I'm curious if I'm wrong about these assumptions:
- It seems like the global difficulty from the main config file increases the likelyhood of a higher-level monster spawner, but doesn't increase the density of spawners?
- The "flags = HARD" setting affects the loot, not the spawners, to represent cases like, how the monsters in the nether dungeon are harder than their overworld counterparts, and warrant better rewards?
- The nearby spawners have something to do with the quality of loot in a chest?

**Comment by BlackJar72** [comment]
The "HARD" setting make the dungeon generate better loot than normal for the mobs -- for dungeons that might be hard for some other reason.

You can sometimes get witches in normal.  The problem is that vanilla has no "elite" level mobs really -- you could find that stuff if you went to the nether and found a wither boss spawner -- yeah, the actual wither, not wither skeletons, which only occurs on nightmare (I think, I don't always remember).  Level 8 loot is rare with elite mobs, common with full bosses, but does appear with weaker mobs.  For Doomlike Dungeons to reach it real potential you basically have to add modded mobs to the theme (or cheese it be decide zombies are now "elite").  This is actually why I've started working on a more "heavy weight" mod -- not everyone uses modded mobs to begin, and not all of those that do want to edit themes to add them.  Unfortunately only I couple mods I know of (Lycanite's mobs and Dungoen Mobs) will actually add mobs on their own, and not in all version (I broke my own API a couple times).

In short, witches (then only level three mob in vanilla, really) may show up in the boss room, but without elite mobs (vanilla has none) or actual bosses you will never find level 8 loot.  So, its modded mobs or multiple withers for that.  Skeleton / zombie / spider rooms will always have weak loot.

**Comment by BlackJar72** [comment]
I don't know when the RPG Dungeons will be available either at this time -- I have a lot of mobs to make, then back to dungeons -- and I've been distracted again by playing with the biome provider (a totally different mod).

**Comment by BlackJar72** [comment]
Oh, now I remember -- I forgot -- it does add enchanted books independently of loot tables.  This was so that regular book could be added as low level loot.

I'll look into this with my own tests and see what I find.

**Comment by BlackJar72** [comment]
OK, I did something similar, and got:

148 heal 1
67 heal 2
12 heal 3

74 gear 1
36 gear 2
9 gear 3

21 loot 7

There is definitely something wrong with the way treasure item are being placed now.  I'll try to look into it in the next few days.

**Comment by TeetoKaziim** [comment]
" For Doomlike Dungeons to reach it real potential you basically have to add modded mobs to the theme (or cheese it be decide zombies are now "elite")."

Yeah, for my tests (configs posted above), this is what I did, in case the mob spawners influenced the loot (which it sounds like they do). My altered theme has:
Zombie common, Cave Spider Hard, Enderman Brute, Creeper Elite, Witch Boss, just for testing purposes. This is not something I'd ship in a real pack, just trying to test as thoroughly as possible.

In an actual pack, you're right, the power level of vanilla mobs, outside the wither and dragon, is pretty flat, so it's hard to justify placing them in the "harder" slots in a theme. Thanks again for the info and for looking into this!  The best to you!

**Comment by BlackJar72** [comment]
OK, I fixed it -- one line of code was messing everything up.  I wish someone had noticed and mentioned it sooner -- apparently this bug has been in the mod since Minecraft 1.7.10!

**Comment by TeetoKaziim** [comment]
I gotta admit, I've been using this mod for like, a year and a half, and, it wasn't until I went REALLY hard on changing the loot pools, that I noticed.

With random stuff, it's always easy to think "oh, well, I just haven't gotten that drop yet", so I can see why folks might not have said anything.

Thanks again!

**Comment by Fabboz** [comment]
with the lastest version for 1.7.10 the chests dont spawn.
version 1.9.2 is working very fine.
no problem with loot even with mods items only level 8 is very rare, i found a wither skull in past but no nether star etc.

**Comment by BlackJar72** [comment]
Are you sure?  I just ran that now and had no problems, plenty of chest.  This exact version:
[http://www.mediafire.com/file/81gxogn9pg1qp01/DoomlikeDungeons-1.10.2-MC1.7.10.jar](http://www.mediafire.com/file/81gxogn9pg1qp01/DoomlikeDungeons-1.10.2-MC1.7.10.jar)

Which is also here:
[https://minecraft.curseforge.com/projects/project-74246/files/2509948](https://minecraft.curseforge.com/projects/project-74246/files/2509948)

That version 1.10.2, with "1.10.2-MC1.7.10" in the title?

If so I have no idea what is going on because it works for me.

Also, are you sure you have the right Forge version -- I've had mods fail because of using an outdated version of Forge before.

(I'm going to close this in the Morning -- your issue is completely different from the one which this was started with -- that issue is solved.)
